snippets/algorithm/maximum_disjoint_subarray_sum.cpp

```diff
@@ -0,0 +1,56 @@
+vector<int> s1, s2;
+// pairs of (total_sum, positive_part)
+vector<pair<int,int>> s1_sums;
+vector<pair<int,int>> s2_sums;
+
+
+// a state is (index, positive part, negative part)
+// so the sum is positive part - negative part
+void find_all_s1_sums(int i, int pos, int neg) {
+    s1_sums.push_back({pos-neg, pos});
+    if(i == s1.size()) return;
+    find_all_s1_sums(i+1, pos, neg);
+    find_all_s1_sums(i+1, pos+s1[i], neg);
+    find_all_s1_sums(i+1, pos, neg+s1[i]);
+}
+
+void find_all_s2_sums(int i, int pos, int neg) {
+    s2_sums.push_back({pos-neg, pos});
+    if(i == s2.size()) return;
+    find_all_s2_sums(i+1, pos, neg);
+    find_all_s2_sums(i+1, pos+s2[i], neg);
+    find_all_s2_sums(i+1, pos, neg+s2[i]);
+}
```

Comment:
can these two methods be combined like:

```c++
find_all_sums(vector<int>& vals, vector<pair<int,int>>& sums, int i, int pos, int neg) {
    sums.push_back({pos-neg, pos});
    if(i == vals.size()) return;
    find_all_sums(vals, sums, i+1, pos, neg);
    find_all_sums(vals, sums, i+1, pos+s2[i], neg);
    find_all_sums(vals, sums, i+1, pos, neg+s2[i]);
}
```

Reply:
and with that the s1,s2,s1_sums,s2_sums, etc. can all be contained within the maximum disjoin subset sum method